### DISCUSSION TITLE
Inline IL

### DISCUSSION THREAD

**Discussion by thatbakamono** [dotnet/csharplang#4395] [Language Ideas]
There are many cases where you cannot just write C# code because some things are unachievable in C# and you have to use IL but there are some obstacles. If you want to use IL, you have to create another project for it and when you have only small parts of your project in IL, it causes code to be a lot harder to navigate because now it's fragmented. I would propose to add inline IL to the language, like C/C++/Rust does with ASM. 

Usages?

For example ``System.Runtime.CompilerServices.Unsafe`` could benefit from this.

How could it look like? (I'm not an expert so there might be a better way to do this)

```cs
public static unsafe bool IsAddressGreaterThan<T>(T left, T right)
{
	return il 
	{
		cgt.un
	}
}

public static unsafe naked bool IsAddressGreaterThan<T>(T left, T right)
{
	il 
	{
		ldarg.0
		ldarg.1
		cgt.un
		ret
	}
}
```

``naked`` means you have to handle everything by yourself, the compiler won't insert anything into the method body.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/226

**Comment by gfoidl** [comment]
> If you want to use IL, you have to create another project for it and when you have only small parts of your project in IL

You could use something like https://github.com/ltrzesniewski/InlineIL.Fody too.

**Reply by thatbakamono** [reply]
Yeah, that library is pretty powerful but it would be awesome to see this as a language feature.

**Comment by svick** [comment]
Are there any use cases for this except for `System.Runtime.CompilerServices.Unsafe`? I don't think it makes sense to add a complicated language feature just to simplify code in a single library that already exists.

**Reply by canton7** [reply]
There are 50,000 downloads of InlineIL.Fody, if that's any indication.

**Reply by JanTamis** [reply]
I agree it makes the language very compilated and i think 50,000 isn't enough.

**Reply by Mrxx99** [reply]
Would also be great for use in Source Generators to produce high performant code. E.g. it could be used for a source generator that compiles Regex.

**Reply by CyrusNajmabadi** [reply]
Darn.  Lost my response.  My general take is:

1. i would actually be ok with allowing this if you have to supply *all* the IL for a method body.
2. I find hte idea of mixing/matching actual C# and IL too worrisome.  I feel that will cause problems in the future, esp. if and when we change emit strategies for normal c# code.
3. i'd like some *very* compelling scenarios for why this would be needed.

Overall though, i don't want this.  I'd prefer we just expose whatever features we need as actual language features if the benefit is there.

**Reply by svick** [reply]
@Mrxx99 Can you explain why exactly would this be useful for source generators? How would it help you make them generate more performant code?

**Reply by PathogenDavid** [reply]
Niche scenario, but something like inline IL could potentially simplify using source generators to precompile regex https://github.com/dotnet/runtime/issues/44676.

One of the big problems with that effort is dealing with the fact the existing regex compiler emits IL but source generators obviously need to emit source code. In theory inline IL could potentially allow re-using most of the existing logic of the current regex-to-IL compiler. (Rather than creating an additional regex-to-C# compiler, which seems to be the current plan.)

**Reply by tannergooding** [reply]
Something else to consider however is that while the JIT can handle all kinds of IL; it is typically optimized around the IL that C#/VB/F# emit.
So just because something looks like good IL, doesn't mean that it will produce good code and ultimately handling things in C# will likely wind up better.
* This is also not just limited to CIL/C#/JIT. LLVM has the same notice for compiler authors about it being optimized for what Clang produces; as does JVM for Java/Kotlin.

**Comment by john-h-k** [comment]
> For example System.Runtime.CompilerServices.Unsafe could benefit from this.

The entire point of that package is that it does things not acheivable in C#

**Reply by thatbakamono** [reply]
Yeah, you are right but it doesn't imply that the idea is wrong or unnecessary.

**Comment by john-h-k** [comment]
>  some things are unachievable in C#

They are basically limited to 2 things now. `ldtoken` and taking the address of a virtual method

**Comment by AartBluestoke** [comment]
Is the relative address order of 2 objects even a stable thing under garbage collection?
my main use for il is generating non-reflection based property assignment operators for a class discovered at runtime.

**Reply by john-h-k** [reply]
@AartBluestoke No, `IsAddress[Less][Greater]Than` is UB across 2 objects. It is only well defined within a single object (between 2 fields or 2 elements of an array)

**Reply by wesnerm** [reply]
Even though it is undefined behavior, it is usable because the value is often stable. This is useful for union-find operations in which a canonical representation is needed for efficiency.

To make equality comparisons fast in a library I created. If I detect two objects are equal, I choose one and replace the references to the other one to point to ones with the lower address. This assumes lower addresses are more likely Gen 2 and less likely to be temporary.

**Comment by wes-sleeman** [comment]
I like the idea of being able to use an `il { … }` block in any `unsafe` context (and ditching the `naked` modifier entirely), but I think it would be more powerful if used more like `asm` block in C/C++ instead of the expression-like syntax shown here. Specifically, it would be great if it could just inject whatever was put in that block directly into the output CIL assembly at the appropriate point.

**Reply by thatbakamono** [reply]
Yeah, I agree. I'm not an IL expert so I proposed both options for other people to review but as I think of it, the expression-like syntax is probably impossible to implement because of the way IL works, so we should just focus on asm-like blocks, they would simplify many things and I would love to see this as a part of the language.

**Reply by msedi** [reply]
I agree. If the few open things with IL are solved, `asm` is maybe the better solution. Especially with things you cannot currently do because you need runtime support (e.g. AVX512). Even though this creates platform dependent code in the sense as it was(is done in C/C++) with #if..#elses I would leave this open to the developer.

**Reply by svick** [reply]
@msedi To me, `asm` seems like it would require significantly more work than continuing expanding hardware intrinsics (https://github.com/dotnet/runtime/issues/35773).

**Reply by msedi** [reply]
@svick: Sure, I agree with you. The question if the team that is expanding the intrinsics has enough time/resources for improvement. As I understood @tannergooding  in the discussion you have linked a lot of additional APIs have to be added.

Adding AVX512 would really help in our scenario. Sure, I can always go and write my interop method in C/C++ but then I might have the performance drop from the interop call and honestly, nowadays I hate configuring VC projects, they are so much more verbose than a CS project ;-).

There is also a some idea in the discussion to automate intrinsics creation so that the "framework" is always filled with the latest intrinsics. For me SVML would be of great use because we have a lot logs and exps in our code. Are the intrinsics generic enough to create an automatic code generation?

**Reply by tannergooding** [reply]
> To me, asm seems like it would require significantly more work than continuing expanding hardware intrinsics

I think the original commenter was comparing the proposed `il` feature to the C++ `asm` feature, not suggesting that C# get the ability to encode C++ like `asm`.
Actually supporting x86 (let alone x64, arm32, arm64, etc) instructions in C# via `asm` would be essentially impossible without a corresponding framework API, in which case you should just use the corresponding framework API (e.g. `System.Runtime.Intrinsics`).

> Adding AVX512 would really help in our scenario

AVX-512 is currently on the general backlog with no definitive implementation date. It comprises somewhere between 3000 and 4500 new methods and is largely only available in server hardware. When you look at something like the [Steam Hardware Survey](https://store.steampowered.com/hwsurvey) (see `Other Settings`) it is only just getting close to broaching 1%. I imagine it will be implemented eventually, but I feel there needs to be both broader market share and clearer status on the usability (it's largely targeted towards HPC and AI/ML, so we couldn't easily update most of the framework to utilize it considering "average" input sizes).

> There is also a some idea in the discussion to automate intrinsics creation so that the "framework" is always filled with the latest intrinsics

Adding new intrinsics to the JIT ranges from trivial (simply adding a new line to a header file table and everything lights up) to vastly complex because the JIT has no support for the scenario.
All of AVX-512 currently falls into the latter category because it involves new registers, a new instruction encoding, a new concept around "masking" which itself requires new registers and encoding support, and a few other bits. Once the initial work to support those new concepts is done, then most remaining intrinsics from AVX-512 should become simply adding entries to the table (and of course taking all 3000+ methods through API review 😄).

> Are the intrinsics generic enough to create an automatic code generation

The intrinsics are platform specific, so you could write a source generator for example that does something like OpenMP, but you would need to handle the platform specific differences between x86 and ARM.

**Reply by msedi** [reply]
@tannergooding: I see your point. What would be your suggestion (just ot make sure: I do not mean this ironically, I'm really interested on what you suggest)? I also agree that there needs to be more market share to justify implementing AVX-512.

For natural reasons .NET is always a bit behind new technologies. In C++ it is much easer to get access to new things:

- CUDA integrates better
- AVX can be used out of the box
- SVML can be used

For every interop technology either it's not possible to include natively in .NET or I have to invest a huge amount of time or money to get access to these technolgies, e.g. 

- OpenGL (OpenTK, SharpGL)
- DirectX (SharpDX,SlimDX)
- Vulkan

Most of these wrappers are then abandoned or not maintained anymore which always withhold from using them at all.

Don't get me wrong, I really love .NET and C# and the infrastructure around it, but using these technologies costs my team way to much work where I can simply **use** the technology in C++ (or Python). For CUDA as an example we wrote our own compilers and interop wrappers just to use CUDA in C# as it can be used in C++. The other option would have been to simply abandon C# and do everything completely in C++ (which was no option for us). 

So my question is if there are better ways or guidelines of how to use and integrate things that are currently not supported by C#?

**Reply by bernd5** [reply]
Clearly those low-level technologies are written in C or C++ and therefore natively available in those languages (which is not true for Python which needs a wrapper, too).

What we have today is the `extern` keyword which can (and is) be used to refer to arbitrary external/native implementations. The problem here is that only for example the corelibrary/runtime can use native implementations without to much interop overhead (FCall, QCall, ...).
I don't think that we have to adjust the language but maybe the toolset to allow "extern" functions without PInvoke/dlls.

And even if we would allow for example embedded IL code, what would we get from it? I mean you would have to define / copy data structures to interact with the native library like we need to do with P/Invoke today. We can't just include a C++ header or something like that.

**Reply by msedi** [reply]
@bernd5: 

> which is not true for Python which needs a wrapper, too 

Sure you are right. It was maybe wrong to mention this in the same context. I assume the community is currently very large to create a lot of bindings.

> And even if we would allow for example embedded IL code, what would we get from it? I mean you would have to define / copy data structures to interact with the native library like we need to do with P/Invoke today. We can't just include a C++ header or something like that.

From my point of view let alone IL to be intgrated into C#. If everything is supported in C# even the last few missing parts I think something like IL weaving is not required anymore.

But, as you have mention something like creating a C# Wrapper from C++ header files would help a lot. For our CUDA implementation we are using CLang, retrieve the syntax tree and create the necessary call to the CUDA Driver API, but this is very specific to CUDA.

We have once tried SWIG but it didn't work so well and we haven't had the time to really deeply investigate. Something out of the box would be very appreciated I think. But this is maybe another discussion and has nothing to do with IL.

**Reply by HaloFour** [reply]
I always thought that this was one of the major roles of C++/CLI, to enable writing the glue code between the unmanaged (and specifically C++) worlds and the managed world.  But I've not followed that language in many years so I don't know if/how it's evolved with .NET Core.

**Reply by bernd5** [reply]
This is still true. And it works in the opposite direction, too. I used it to create Python bindings to CSharp / DotNet code. But C++/CLI is currently only supported on Windows AFAIK.

**Reply by msedi** [reply]
@bernd5 + @HaloFour: Sure, but not having to write the C++/CLI Wrapper would of course be pretty helpful. Something like the old COM had it's TLBs so that .NET knows a way how to interpret the TLB.

**Reply by HaloFour** [reply]
TLBs were designed very specifically for that purpose.  C++ headers not so much.

**Reply by msedi** [reply]
@HaloFour: Yes that makes sense. I'M a lazy guy I just want to optimize my efforts wherever possible. And currently I see in my team that the efforts to write wrappers is extremely high, where the DLLs are not only accessible by simple P/Invokes but need C++/CLI wrappers which is of course also very error prone.

Honestly while discussing, I came again across another project: SharpCppGen. Obviously a way exists, but I don't know it works how good and how much effort the team has put into this.

**Reply by HaloFour** [reply]
I'd imagine that swig is probably the most mature of those tools as it's been around a while and is designed to work with a number of languages.  I think interpreting C++ headers like that will always be incredibly finicky and require human intervention in all but the simpler cases.

**Reply by tannergooding** [reply]
You might want to check out tools like http://github.com/microsoft/clangsharp which support taking in a C header file and generating 1-to-1 C# bindings.

This is what is currently used behind the scenes for https://github.com/microsoft/win32metadata and what I use for my own bindings such as http://github.com/terrafx/terrafx.interop.windows (User32, DX10, DX11, DX12, D2D1, DWrite, WIC, etc), http://github.com/terrafx/terrafx.interop.vulkan, http://github.com/terrafx/terrafx.interop.xlib, etc.

It still falls short on certain C++ specific constructs, but C++ code in general isn't recommended to bind against due to cross plat differences and possible ABI breaks across compiler versions.
(and its open source so anyone is free to help handle new scenarios that are important if they come up; plus I'm the maintainer so I'd appreciate it 😄)

**Comment by bernd5** [comment]
I would prefer that we add the missing parts simply to the language (like we did with function pointers) - there is no need for inline assembly or something similar!

**Comment by tannergooding** [comment]
I commented on a couple of other threads but in general I wouldn't be in favor of such a feature.

Even in C/C++, the `asm` feature is not standard and the support for it can greatly vary (x64 MSVC has no support). When it does exist, you basically want a naked function or nothing at all because it can become very hard for the compiler to actually rationalize what the method is doing and to actually optimize for it, so in many cases you end up with poorer optimizations (or often none at all). GCC probably has the best support, but it also requires you too write pseudo-assembly and provide many additional annotations on the `asm` block to tell the compiler exactly what the inputs, outputs, trashed registers, and other requirements are.

In C#, this would tie .NET to IL where it (strictly speaking) has no such requirement today. It would likely require formalizing IL as part of the actual language which to me just seems like general badness. Likewise, I think we are down to less than 10 IL instructions that can't actually be represented by C# today and they largely revolve around metadata tokens and areas the JIT has limited interaction with today (meaning it may not even optimize it like you'd hope). Where these missing cases are actually useful, it would probably be beneficial to try to get proper language support where appropriate.

For the items that are more esoteric, it is likely much simpler to get a proposal up for a new framework API in `System.Runtime.CompilerServices.Unsafe` (which is full of things C# can't directly do today; some of them for good reason) or via another more appropriate setup. It may even be simpler for the compiler (not the language) to recognize some of those methods as `intrinsic` and to emit the relevant IL for them (this was discussed on another thread related to the IL linker, it just "locks" the implementation of that method to never being changed, functionally speaking).

**Reply by HaloFour** [reply]
_cough_ `methodof(...)` _cough_ 😁

**Reply by TahirAhmadov** [reply]
In FOOF we trust
